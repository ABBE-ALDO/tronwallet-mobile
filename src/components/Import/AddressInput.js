import React, { PureComponent, Fragment } from 'react'
import { func, string } from 'prop-types'

import tl from '../../utils/i18n'
import { isAddressValid } from '../../services/address'

import Input from '../Input'
import { Text } from '../Utils'
import { Colors } from '../DesignSystem'

import InputOptions from './InputOptions'
import QRScannerModal from './QRScannerModal'

class AddressInput extends PureComponent {
  static propTypes = {
    onChangeText: func,
    innerRef: func,
    onSubmitEditing: func,
    labelWrapperColor: string
  }

  static defaultProps = {
    onChangeText: () => {},
    innerRef: () => {},
    onSubmitEditing: () => {},
    labelWrapperColor: Colors.background
  }

  state = {
    address: '',
    addressError: null,
    qrModalVisible: false
  }

  _changeAddress = (address) => {
    const addressError = isAddressValid(address) ? null : tl.t('send.error.incompleteAddress')

    const { onChangeText } = this.props
    onChangeText(address, addressError)

    this.setState({ address, addressError })
  }

  _openModal = () => this.setState({ qrModalVisible: true })

  _closeModal = () => this.setState({ qrModalVisible: false })

  _onReadScanner = ({ data }) => {
    this._changeAddress(data)
    this._closeModal()
  }

  _rightContent = () => (
    <InputOptions
      onPaste={(content) => this._changeAddress(content)}
      onPressScanner={this._openModal}
    />
  )

  render () {
    const { innerRef, onSubmitEditing, labelWrapperColor } = this.props
    const { address, addressError, qrModalVisible } = this.state

    return (
      <Fragment>
        <Input
          innerRef={innerRef}
          label={tl.t('address').toUpperCase()}
          labelWrapperColor={labelWrapperColor}
          rightContent={this._rightContent}
          value={address}
          onChangeText={to => this._changeAddress(to)}
          onSubmitEditing={onSubmitEditing}
          placeholder=' '
        />
        {!!addressError && (
          <Text marginY={8} size='xsmall' color={Colors.redError}>
            {addressError}
          </Text>
        )}
        <QRScannerModal
          visible={qrModalVisible}
          onClose={this._closeModal}
          onRead={this._changeAddress}
        />
      </Fragment>
    )
  }
}

export default AddressInput
