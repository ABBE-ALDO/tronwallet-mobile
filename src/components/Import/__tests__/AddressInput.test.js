import React from 'react'
import { shallow } from 'enzyme'

import AddressInput from '../AddressInput'

jest.mock('../../../utils/i18n')

describe('Address input component', () => {
  const setup = propOverrides => {
    const props = Object.assign({
    }, propOverrides)

    const wrapper = shallow(
      <AddressInput {...props} />
    )

    return {
      props,
      wrapper
    }
  }

  test('Should match when render component', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })
})
