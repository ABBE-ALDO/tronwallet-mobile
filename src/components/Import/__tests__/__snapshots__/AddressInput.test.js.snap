// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address input component Should match when render component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddressInput
    innerRef={[Function]}
    labelWrapperColor="#191A2A"
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Input
          borderColor="#51526B"
          editable={true}
          innerRef={[Function]}
          label="ADDRESS"
          labelWrapperColor="#191A2A"
          numbersOnly={false}
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder=" "
          placeholderTextColor="#66688F"
          returnKeyType="send"
          rightContent={[Function]}
          showPlaceholder={true}
          value=""
        />,
        false,
        <QRScannerModal
          onClose={[Function]}
          onRead={[Function]}
          visible={false}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "borderColor": "#51526B",
          "editable": true,
          "innerRef": [Function],
          "label": "ADDRESS",
          "labelWrapperColor": "#191A2A",
          "numbersOnly": false,
          "onChangeText": [Function],
          "onSubmitEditing": [Function],
          "placeholder": " ",
          "placeholderTextColor": "#66688F",
          "returnKeyType": "send",
          "rightContent": [Function],
          "showPlaceholder": true,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onClose": [Function],
          "onRead": [Function],
          "visible": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Input
            borderColor="#51526B"
            editable={true}
            innerRef={[Function]}
            label="ADDRESS"
            labelWrapperColor="#191A2A"
            numbersOnly={false}
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            placeholder=" "
            placeholderTextColor="#66688F"
            returnKeyType="send"
            rightContent={[Function]}
            showPlaceholder={true}
            value=""
          />,
          false,
          <QRScannerModal
            onClose={[Function]}
            onRead={[Function]}
            visible={false}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "borderColor": "#51526B",
            "editable": true,
            "innerRef": [Function],
            "label": "ADDRESS",
            "labelWrapperColor": "#191A2A",
            "numbersOnly": false,
            "onChangeText": [Function],
            "onSubmitEditing": [Function],
            "placeholder": " ",
            "placeholderTextColor": "#66688F",
            "returnKeyType": "send",
            "rightContent": [Function],
            "showPlaceholder": true,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onClose": [Function],
            "onRead": [Function],
            "visible": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenThreeAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
